### ISSUE TITLE
Memory leak during product export

### ISSUE THREAD

**Issue by mathewrapid** [akeneo/pim-community-dev#4147]
I'm using the EnhancedConnectorBundle to export all the products that are in the system regardless their status, completeness etc.

There are almost 110 000 products in total. For every 100 products the memory consumption increases by ~30 MB. There must be a memory leak somewhere within the while loop in https://github.com/akeneo/pim-community-dev/blob/v1.5.0/src/Akeneo/Component/Batch/Step/ItemStep.php#L174

Link to [Blackfire](https://blackfire.io/profiles/d34359a8-f052-466d-8d4f-9b2fad7e6b85/graph) graph.

The export worked with the previous installed version 
[akeneo-pim-system-info_2016-02-29_12-15.txt](https://github.com/akeneo/pim-community-dev/files/172028/akeneo-pim-system-info_2016-02-29_12-15.txt)

Current system info: [akeneo-pim-system-info_2016-03-14_15-38.txt](https://github.com/akeneo/pim-community-dev/files/172027/akeneo-pim-system-info_2016-03-14_15-38.txt)

**Comment by grena** [comment]
Hello @mathewrapid !
Thank you for reporting this kind of behaviour, we really try to work hard on scalability issues.

Sorry for asking, but did you launch your export with a prod environment?
I'm pinging the EnhancedConnector team to see if they know something about it.

Kind regards,

**Comment by mathewrapid** [comment]
Yes, the command is as follows `php app/console akeneo:batch:job csv_all_products_export --env=prod`
I'll try to reproduce this with a v1.4.20 installation and report my findings later.

**Comment by grena** [comment]
Thank you for keeping us up to date! :+1: 
Do you have a lot of media attributes to export?

**Comment by mathewrapid** [comment]
Media attributes: none

**Comment by mathewrapid** [comment]
Tested with v1.4.20.

Same MySQL database, same MongoDB, same export, huge difference.

`echo (memory_get_usage(true) / 1024 / 1024) . PHP_EOL;`

v1.5.0 (batch size 10) `Akeneo\Component\Batch\Step\ItemStep::doExecute`

Iteration|Memory usage in MB
------------|------------
1|60.5
2|63.75
3|66.5
4|70.25
5|73
6|75.75
7|78.25
8|81
9|85.5
10|88.25

v1.4.20 (batch size 10) `Akeneo\Bundle\BatchBundle\Step\ItemStep::doExecute`

Iteration|Memory usage in MB
------------|------------
1|59.5
2|60.5
3|60.75
4|60.75
5|61.5
6|61.5
7|61.5
8|61.5
9|61.5
10|61.5
11|65
12|65
13|65
14|65
15|65
16|65
17|65
18|65
19|65
20|65
21|65
22|65
23|65
24|65
25|65
26|65
27|65
28|65
29|65
30|65
31|65
32|65
33|65
34|65
35|65
36|65
37|65
38|65
39|65
40|65
41|65
42|65
43|65
44|65
45|65
46|65

**Comment by mathewrapid** [comment]
This might not be exclusive to the EnhancedConnector export, the server runs out of memory even for imports with v1.5.0

**Comment by solivier** [comment]
Hi @mathewrapid,

Thanks a lot for those reports. Did you bench it on a 1.5.0 without any external bundle ?

cheers,
Olivier

**Comment by mathewrapid** [comment]
Same custom codebase with both tests, so no.

My v1.4.20 installation uses `doctrine/mongodb-odm (1.0.4)` where the v1.5.0 uses the `1.0.5` version of that. https://github.com/doctrine/mongodb-odm/compare/1.0.4...1.0.5
Don't know if that could have anything to do with this.

**Comment by Hocdoc** [comment]
+1 I'm also interested in this issue.

As a side note my first tests with the new version: 
- For our use case the export for a small number of products is a lot faster in Akeneo 1.5 than in Akeneo 1.4. 
- Unfortunately the export of >1000 products gets much slower (or are not possible) in Akeneo 1.4 compared with Akeneo 1.5.

To compare I've used the same data for both Versions while using the MySQL-only data store.

**Comment by nidup** [comment]
Hi @mathewrapid 

Thank you a lot for reporting this issue!

About MongoDB dependencies, in 1.4 we use,
```
        "doctrine/mongodb-odm-bundle": "v3.0.0-BETA6",
        "doctrine/mongodb-odm": "v1.0.0-BETA12@dev",
```
In 1.5, we use,
```
        "doctrine/mongodb-odm-bundle": "3.0.1",
``` 

The mongodb-odm v1.0.0-BETA12 contains a memory leak issue related to cascade detach, we fixed it in 1.4 with https://github.com/akeneo/pim-community-dev/blob/master/src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/Common/Detacher/ObjectDetacher.php#L186

When we manipulate bulk of products (import, export, mass edit, etc), we load them page by page, we process and we detach them from the Doctrine UOW, if we don't do that, all products stay in memory and consume more an more memory with each bulk page (we use our ObjectDetacher to do so).

During a bulk process,
 - all objects belonging to a product as product values, metric, price are detached once loaded/processed
 - the objects linked to the product but shared with other products as categories, families, attributes, attribute options, etc always stay in memory (to avoid to fetch them from database for each product page)

When, we run a bulk process, it's normal to see the memory usage increase with each page, it's related to the fetching of new linked objects in memory (attributes, categories, etc). Once the linked data used by the whole bulk are loaded in memory, the memory stays stable (the garbage collector cleans products not anymore referenced by the UOW but not the linked data still referenced).

Your 1.4 profiling clearly demonstrates this, memory increases a bit with each page of products (fetching new linked data), then stabilize at 65 MB for the whole process. For your 1.5 profiling i would see more product pages to know when/if the memory stabilizes at some point? Does the process stops due to reached memory limit at some point ?

About memory usage, 80 MB does not look crazy to me compared with other data i see on projects, when having a bunch of linked data (10k attributes, category, etc), the memory can stabilize around 500 MB. We advise to deserve 768 MB for PHP CLI cf http://docs.akeneo.com/latest/reference/technical_information/server_side_setup_for_hosting.html

If the memory never stabilizes it may indicates a memory leak that we could have introduced in a 1.5 feature or through a 1.5 changed dependency. BTW, we added several performances build on the application in 1.5, still in in early stage but they allow us to check that the memory does not reach a defined limit for one of our defined catalog during an import or an export for instance.

Your blackfire.io profile shows that a lot of time is spent to convert metric and add missing price before to export, if confirme, the memleak could come from these.

About scalability and performance, you may be interested by,
 - http://docs.akeneo.com/latest/reference/scalability_guide/index.html
 - http://docs.akeneo.com/latest/reference/performances_guide/index.html

Recently, we also did tests with PHP7 https://medium.com/akeneo-labs/akeneo-pim-1-5-and-the-php-7-support-of-the-cli-fa5a87325438#.v3ykhvoaq 

@Hocdoc can you provide same kind of data table than @mathewrapid with product page an used memory please? BTW, I seen your recent PRs :tada: ! thanks a bunch for your help, we'll quickly merge!

**Comment by mathewrapid** [comment]
@nidup No, 80 MB is not much, but the memory usage never stabilizes. I did run out of 3 GB memory with that export on 1.5. With 1.4.20 it eventually stabilized at approx 100 MB for the same export.

**Comment by nidup** [comment]
@mathewrapid thank your very much for the details, we're on the same page :smile: I agree that 3 GB is way toooo much, i'm testing with our different default data sets and i keep you posted.

**Comment by nidup** [comment]
@mathewrapid I tried to reproduce with our tests data set and the used memory stabilizes at ~65MB for a product export (1.5 standard installation with mongodb 2.4, php 5.5).

I guess your data set highlights a memory leak that our own data sets does not allow to reproduce.
I'm very interesting to find and fix it, are you agree to share your project code and database dump to help me to reproduce and dive into this issue? If ok, could you send me an email at nicolas at akeneo dot com with accesses to these data and the code of the job that need to be run?

Thank you!

**Comment by mathewrapid** [comment]
I need to check that with the customer if they'll agree to handing over the data for testing.

**Comment by nidup** [comment]
Sure, no pb. A second possibility is to anonymize your catalog data. A third possibility is to generate a fake catalog with https://github.com/akeneo-labs/DataGeneratorBundle. Using same properties than your real catalog to be able to reproduce the issue. 1st is the more relevant, then 2nd then 3rd.

**Comment by tebinu** [comment]
The problem still exists. The problems root is somewhere in the reader 
\Akeneo\Component\Batch\Step\ItemStep -> $readItem = $this->reader->read(); 

We will send db dumps

**Comment by mathewrapid** [comment]
Just saw this https://github.com/akeneo/pim-community-dev/blob/1.5/CHANGELOG-1.5.md#154-2016-06-01

We'll test the export with the 1.5.4 and see if it fixes the issue.

**Comment by snc** [comment]
Has this fixed your issue @mathewrapid? I'm facing memory leaks on 1.6.2 on a custom export.

**Comment by mathewrapid** [comment]
Hi @snc,

For now we've stuck with the 1.4.x branch. Sorry to say that we not made any progress with this yet.

**Comment by nei** [comment]
I've updated from 1.5.8 to 1.5.20 and the product export is not finishing properly.

I am trying to export just one channel with 17207 products. 

I've noticed that on about 4470 items processed the memory is on 503Mb. I can see it growing about 1Mb each 10 items processed.

**Comment by nidup** [comment]
Hi @nei 

Usually during product export, the memory consumption grows with each bulk (because fetch attributes,families, options, categories in memory) then stabilize after few bulks (around 200Mb, 300 Mb or more depending on the complexity of the catalog and the typology of products).

In your case, when exporting with the exact same data set with v1.5.8, it stabilizes around which amount of memory and how much time it takes to finish? Same question with v1.5.20? I guess with your message that it does not finished at all, does it raised the max memory limit?

**Comment by nei** [comment]
On 1.5.8 It was successfully importing 13145 in 51 min but we were not measuring the memory usage. was setup for unlimited. 

Now, it is taking 2 hrs to run 17k and then it shows an error without much information.

![screen shot 2017-03-21 at 3 52 38 pm](https://cloud.githubusercontent.com/assets/33385/24156474/85fb558e-0e4e-11e7-9273-f84a6fee4e48.png)

Now I am trying to run locally and it took 3 hrs and imported just 8k products and then I noticed the memory keep going up. I will do some more benchmarking.

**Comment by nei** [comment]
I discovered that on production it was crashing because the memory usage.

Attached the timespan and memory usage data below.

This is the latest 1.5.20. I did not tested on 1.58 though but the issue was probably there already we just noticed now because the dataset increased a little bit.

I reproduced the issue locally. it took 14 hours to finish 12k items and the memory never stabilizes.
[product_export-perf.xlsx](https://github.com/akeneo/pim-community-dev/files/861002/product_export-perf.xlsx)

**Comment by nidup** [comment]
Hi, looks like a small memory leak, it may come from the data set or custom code.

I'm interesting to see the blackfireio profiles or reports from https://github.com/BitOne/php-meminfo

**Comment by oliverde8** [comment]
Hi, 

I also had a few memory issues, lately I exported & imported 20K products for a customer, from a Akeneo 1.5 to a Akeneo 1.6 .12. Both Akeneo's went out of memory. I had to increase the memory limit to 6gb to make the imports/export work. 

I suspect the issue is coming from linked products, the catalog has alot of configurable & also linked products (crossell & upsell) and after poking in the code I couldn't find where those "related" products were detached. 

I hope it can help. I had very little time & therefore couldn't investigate anymore then that.

**Comment by nidup** [comment]
Hi @oliverde8

6 GB is huged, did you run the import and export in prod mode? (in dev mode, symfony profiler keeps references on all objects and can cause such a leak).

**Comment by oliverde8** [comment]
Hi @nidup 

Yes I ran it prod mode. It's import that really used alot of memory and as said I am pretty sure the memory leaks comes from the linked products. Each product was linked to at-least 2 other products (beside some basic products).

**Comment by kanduvisla** [comment]
Not sure if this is the right thread, but I'm trying to export 16k products in Akeneo 2.1.4, and I'm also getting the `Allowed memory size of 2147483648 bytes exhausted`-error.

I have media attributes, but they are empty (clean install), and media export is disabled.

The class in question in the error is `vendor/doctrine/dbal/lib/Doctrine/DBAL/Types/JsonArrayType.php`. Could it be that Doctrine is the cause of my problems in this case?

I also looked at https://docs.akeneo.com/2.0/maintain_pim/scalability_guide/more_than_100k_products_to_export.html, where they state that they export a stream so the only limit is your hard disk drive, but clearly this is not the case here.

**Comment by grena** [comment]
Hello @kanduvisla , don't hesitate to open another issue for this behavior, as you run on a pretty recent version :)
